pageserver/src/tenant/timeline.rs

```diff
@@ -2747,16 +2856,35 @@ impl Timeline {
             end: deltas_to_compact.last().unwrap().get_lsn_range().end,
         };
 
+        let remotes = deltas_to_compact
+            .iter()
+            .filter(|l| l.is_remote_layer())
+            .inspect(|l| info!("compact requires download of {}", l.filename().file_name()))
+            .map(|l| {
+                l.clone()
+                    .downcast_remote_layer()
+                    .expect("just checked it is remote layer")
+            })
+            .collect::<Vec<_>>();
+
+        if !remotes.is_empty() {
+            // caller is holding the lock to layer_removal_cs, and we don't want to download while
+            // holding that; in future download_remote_layer might take it as well.
+            return Err(CompactionError::DownloadRequired(remotes));
```

Comment:
Just noticed that we are in fact holding the lock while image creation downloads the layers. Should it be noted here, or alternative approach be taken still (download here)?

Reply:
Noted in [1af4b79](https://github.com/neondatabase/neon/pull/3598/commits/1af4b79d5895e237e6faf8fbc839ce6ea1b5555a)